geode-junit/src/main/java/org/apache/geode/test/junit/rules/gfsh/GfshExecutor.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work for additional information regarding
+ * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package org.apache.geode.test.junit.rules.gfsh;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeoutException;
+
+public interface GfshExecutor {
```

Comment:
Would it be reasonable to extract the common bits from `GfshRule` and `GfshContext` into this interface as default implementations? (This may not be worth the effort if we intend to try the rule/extension injection technique.)

Reply:
`GfshScript` also have several `execute` methods but with different signatures.

The implementations of each rule's `execute` methods are too different to have any common code that could be extracted. However, all `execute` calls do eventually route to a single `execute` method in `GfshContext` that does all of the actual work. The "in-betweeners" just add in more parameters from fields or convert certain parameters such as `File` vs `Path`.